README.rst
=========================
roman_numerals_webservice
=========================


.. image:: https://img.shields.io/pypi/v/roman_numerals_webservice.svg
        :target: https://pypi.python.org/pypi/roman_numerals_webservice

.. image:: https://img.shields.io/travis/DerThorsten/roman_numerals_webservice.svg
        :target: https://travis-ci.org/DerThorsten/roman_numerals_webservice

.. image:: https://readthedocs.org/projects/roman-numerals-webservice/badge/?version=latest
        :target: https://roman-numerals-webservice.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/DerThorsten/roman_numerals_webservice/shield.svg
        :target: https://pyup.io/repos/github/DerThorsten/roman_numerals_webservice/
        :alt: Updates

.. image:: https://img.shields.io/badge/docker-ready-blue.svg
        :target: https://hub.docker.com/r/derthorsten/roman_numerals_webservice


A python package for a minimalistic web-service to convert roman numerals to arabic numerals and vice versa.


* Free software: MIT license
* Documentation: https://roman-numerals-webservice.readthedocs.io.

Features
--------

* Python code to to convert Roman numerals to Arabic numerals and vice versa
* Minimalistic web-service to convert Roman numerals to Arabic numerals and vice versa
* Heavy tested
* Nightly builds on `continous integration servers`_
* available as `pypi-package`_ package
* Documentation on `readthedocs.org`_
* Available on `dockerhub`_


.. _`continous integration servers`: https://travis-ci.org/DerThorsten/roman_numerals_webservice
.. _`dockerhub`: https://hub.docker.com/r/derthorsten/roman_numerals_webservice
.. _`readthedocs.org`: https://roman-numerals-webservice.readthedocs.io
.. _`pypi-package`: https://pypi.python.org/pypi/roman_numerals_webservice
.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Try package online`: https://mybinder.org/v2/gh/DerThorsten/roman_numerals_webservice/master?filepath=example_notebooks/Usage.ipynb

HISTORY.rst
=======
History
=======

0.4.0 (2019-07-24)
------------------

* improved documentation
* added nightly builds on travis
* renamed int_to_x to arabic_to_x
* added better notebook
* added Dockerfile

0.3.3 (2019-07-24)
------------------

* changed entry point

0.3.2 (2019-07-23)
------------------

* fix pyi version


0.3.1 (2019-07-23)
------------------

* fixed version issue

0.3.0 (2019-07-23)
------------------

* fixed missing subpackages
* added examples


0.2.0 (2019-07-23)
------------------

* First release on PyPI. 


0.1.0 (2019-07-23)
------------------

* Initial version 


tests/test_roman_numerals_webservice.py
import pytest
import numpy
from roman_numerals_webservice import RomanNumeralsWebservice
from roman_numerals_webservice.roman_numerals import (arabic_to_roman,roman_to_arabic)

from http import HTTPStatus
import cherrypy
import requests
import json

from contextlib import contextmanager
 
@contextmanager
def run_server():
    cherrypy.engine.start()
    cherrypy.engine.wait(cherrypy.engine.states.STARTED)
    yield
    cherrypy.engine.exit()
    cherrypy.engine.block()


def roman_numer_service_runner(func):
    def func_wrapper(self):
        cherrypy.tree.mount(RomanNumeralsWebservice())
        with run_server():
            func(self)
    return func_wrapper


class TestRomanNumeralsWebservice(object):
    url_arabic_to_roman = 'http://localhost:8080/arabic_to_roman'
    url_roman_to_arabic = 'http://localhost:8080/roman_to_arabic'


    @roman_numer_service_runner
    def test_arabic_to_roman_valid_input(self):

        url = TestRomanNumeralsWebservice.url_arabic_to_roman

        # testing all is to slow
        for arabic in list(range(1,4000, 17)) + [3999]:
            r = requests.post(url, json={"arabic" : arabic})
            assert r.status_code == HTTPStatus.OK
            json_res = json.loads(r.json())
            assert json_res["roman"] == arabic_to_roman(arabic)    

    @roman_numer_service_runner
    def test_arabic_to_roman_valid_input_advanced(self):

        url = TestRomanNumeralsWebservice.url_arabic_to_roman

        # in python bool is subclass of int
        r = requests.post(url, json={"arabic" : True})
        json_res = json.loads(r.json())
        assert json_res["roman"] == arabic_to_roman(1)

    @roman_numer_service_runner
    def test_arabic_to_roman_invalid_input(self):

        url = TestRomanNumeralsWebservice.url_arabic_to_roman


        # wrong range
        r = requests.post(url, json={"arabic" : 0})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"arabic" : 4000})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"arabic" : -1})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"arabic" : False})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        # wrong type
        r = requests.post(url, json={"arabic" : '1'})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"arabic" : [1]})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        # wrong key
        r = requests.post(url, json={"Arabic" : 1})
        assert r.status_code == HTTPStatus.BAD_REQUEST
        
        r = requests.post(url, json={"roman" : 1})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={" arabic" : 1})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"arabic " : 1})
        assert r.status_code == HTTPStatus.BAD_REQUEST

    @roman_numer_service_runner
    def test_araic_to_roman_valid_input(self):

        url = TestRomanNumeralsWebservice.url_roman_to_arabic

        # testing all is to slow
        for arabic in list(range(1,4000, 17)) + [3999]:
            roman = arabic_to_roman(arabic)
            r = requests.post(url, json={"roman" : roman})
            assert r.status_code == HTTPStatus.OK
            json_res = json.loads(r.json())
            assert json_res["arabic"] == arabic

        # testing all is to slow
        for arabic in list(range(1,4000, 17)) + [3999]:
            roman = arabic_to_roman(arabic)
            r = requests.post(url, json={"roman" : roman.lower()})
            assert r.status_code == HTTPStatus.OK
            json_res = json.loads(r.json())
            assert json_res["arabic"] == arabic


    @roman_numer_service_runner
    def test_araic_to_invalid_input(self):

        url = TestRomanNumeralsWebservice.url_roman_to_arabic

 
        # wrong whitespace
        r = requests.post(url, json={"roman" : " X"})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        # wrong whitespace
        r = requests.post(url, json={"roman" : "X "})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        # wrong type
        r = requests.post(url, json={"roman" : 1})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"roman" : True})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"roman" : ['X']})
        assert r.status_code == HTTPStatus.BAD_REQUEST    
    
        r = requests.post(url, json={"roman" : ('X',)})
        assert r.status_code == HTTPStatus.BAD_REQUEST    

        # invalid roman numeral
        r = requests.post(url, json={"roman" : 'XXXX'})
        assert r.status_code == HTTPStatus.BAD_REQUEST 

        # not a roman numeral
        r = requests.post(url, json={"roman" : 'IVMCX'})
        assert r.status_code == HTTPStatus.BAD_REQUEST  

        # not a roman numeral
        r = requests.post(url, json={"roman" : 'abc'})
        assert r.status_code == HTTPStatus.BAD_REQUEST 

        # wrong key
        r = requests.post(url, json={" roman" : 'I'})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"roman " : 'I'})
        assert r.status_code == HTTPStatus.BAD_REQUEST  

        r = requests.post(url, json={"arabic" : 'I'})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        r = requests.post(url, json={"Roman5" : 'I'})
        assert r.status_code == HTTPStatus.BAD_REQUEST

        

roman_numerals_webservice/roman_numerals_webservice.py
# -*- coding: utf-8 -*-

"""Main module."""

import cherrypy
import json
from http import HTTPStatus

from . roman_numerals import arabic_to_roman, roman_to_arabic




class RomanNumeralsWebservice(object):


    @cherrypy.expose
    @cherrypy.tools.json_out()
    @cherrypy.tools.json_in()
    @cherrypy.tools.accept(media='application/json')
    def roman_to_arabic(self):
        """Implements endpoint for Roman to Arabic web-service
        
        This method expects json request with a json payload of
        the following form:

        .. code-block:: python

            '{"roman": <input_str>}'

        Where :code:`<input_str>` is the input Roman numerals as string.
        The string can be in CAPTIAL lower or mIxeD case.
        
        
        With curl, one can use this endpoint in the following way:

        .. code-block:: shell
        
            $ curl -d '{"roman" : "XL"}' -H "Content-Type: application/json" -X POST http://localhost:8080/roman_to_arabic
        
        The output will be:

        .. code-block:: shell
            
            {\"arabic\": 40}"
      


        Returns:
            str: json string of the form :code:`'{"arabic": <res_int>}'` where
                :code:`<res_int>` is an integer which represents the 
                    input Roman numeral converted to an Arabic Numeral.

        
        Raises:
            cherrypy.HTTPError: If the input it not a valid Roman numeral
                or if the json payload is ill-formed, a :code:`cherrypy.HTTPError(status=400)` 
                is raised. This error will translate to a BAD_REQUEST HTML status code.
        """

        json_dict = cherrypy.request.json

        # get input
        try:
            roman = json_dict['roman']
        except KeyError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST,
                message='json key "roman" is missing')

        # process input
        try:
            arabic = roman_to_arabic(roman)
        except ValueError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST, 
                message='input is not a canonical roman numeral')
        except TypeError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST, 
                message='input value needs to be a string')

        # return result
        return json.dumps(dict(arabic=arabic))

    @cherrypy.expose
    @cherrypy.tools.json_out()
    @cherrypy.tools.json_in()
    @cherrypy.tools.accept(media='application/json')
    def arabic_to_roman(self):
        """Implements endpoint for Arabic to Roman web-service
        
        This method expects json request with a json payload of
        the following form:

        .. code-block:: python

            '{"roman": <input_str>}'

        Where :code:`<input_str>` is the input Arabic numerals as integer.
        
        
        With curl, one can use this endpoint in the following way:

        .. code-block:: shell
        
            $ curl -d '{"arabic" : 1987}' -H "Content-Type: application/json" -X POST http://localhost:8080/arabic_to_roman
        
        The output will be:
        
        .. code-block:: shell
            
            "{\"roman\": \"MCMLXXXVII\"}"
      


        Returns:
            str: json string of the form :code:`'{"roman": <res_str>}'` where
                :code:`<res_str>` is an str which represents the 
                    input Arabic numeral converted to a Roman Numeral.

        
        Raises:
            cherrypy.HTTPError: If the input is not a in the range :code:`[1,...399]`,
                if the json payload is ill-formed, a :code:`cherrypy.HTTPError(status=400)` 
                is raised. This error will translate to a BAD_REQUEST HTML status code.
        """
        json_dict = cherrypy.request.json
        
        # get input
        try:
            arabic = json_dict['arabic']
        except KeyError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST,
                message='json key "arabic" is missing')

        # process input
        try:
            roman = arabic_to_roman(arabic)

        except TypeError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST, 
                message='input value needs to be an integer')
        except ValueError:
            raise cherrypy.HTTPError(status=HTTPStatus.BAD_REQUEST, 
                message='arabic number must be between 1 and 3999')

        
        return json.dumps(dict(roman=roman))





roman_numerals_webservice/roman_numerals/roman_to_arabic.py
from .arabic_to_roman import arabic_to_roman


class RomanToArabic(object):
    _char_2_number = {"M": 1000, "D": 500, "C": 100, "L": 50, "X": 10, "V": 5, "I": 1}

    @staticmethod
    def convert(roman: str) -> int:
        """ Convert a Roman numeral to an Arabic Numeral.


        To convert Arabic numerals we chose the algorithm from Paul M. Winkler
        presented in  :code:`"Python Cookbook" by David Ascher, Alex Martelli ISBN: 0596001673`.
        since it is arguably the most readable algorithm.
        
        Args:
            roman (str): Roman numeral represented as string.
        
        Raises:
            TypeError: roman is not a string
            ValueError: roman is not a valid Roman numeral

        Returns:
            int : int encoding the input as Arabic numeral       
        """
        if not isinstance(roman, str):
            raise TypeError("expected string, got {}".format(type(roman)))
        roman = roman.upper()
        nums = RomanToArabic._char_2_number
        sum = 0
        for i in range(len(roman)):
            try:
                value = nums[roman[i]]
                # If the next place holds a larger number, this value is negative
                if i + 1 < len(roman) and nums[roman[i + 1]] > value:
                    sum -= value
                else:
                    sum += value
            except KeyError:
                raise ValueError("roman is not a valid Roman numeral: {}".format(roman))
        print("arabic_to_roman(sum)",arabic_to_roman(sum),"roman",roman)
        # easiest test for validity...
        if arabic_to_roman(sum) == roman:
            return sum
        else:
            raise ValueError("roman is not a valid Roman numeral: {}".format(roman))




def roman_to_arabic(arabic: int) -> str:
    """ Convert a Roman numeral to an Arabic Numeral.
    
     Shorthand for :py:meth:`RomanToArabic.convert`, see
    :py:meth:`RomanToArabic.convert` for full documentation.
    
    Args:
        roman (str): Roman numeral represented as string.
    
    Raises:
        TypeError: roman is not a string
        ValueError: roman is not a valid Roman numeral

    Returns:
        int : int encoding the input as Arabic numeral       
    """
    return RomanToArabic.convert(arabic)

roman_numerals_webservice/roman_numerals/arabic_to_roman.py
import numbers

class ArabicToRoman(object):

    _int_nums = ((1000, 'M'), (900, 'CM'), (500, 'D'), 
                 (400, 'CD'), (100, 'C'), (90, 'XC'), 
                 (50, 'L'), (40, 'XL'), (10, 'X'),
                 (9, 'IX'), (5, 'V'), (4, 'IV'), 
                 (1, 'I'))

    @staticmethod
    def convert(arabic : int) -> str:
        """Convert an Arabic numeral to a Roman numeral

        To convert Arabic numerals we chose the algorithm from Paul M. Winkler
        presented in  :code:`"Python Cookbook" by David Ascher, Alex Martelli ISBN: 0596001673`.
        since it is arguably the most readable algorithm.
        
        
        Args:
            arabic (int): Arabic numeral represented as integer. The number must be be in :code:`[1,...,3999]`
        
        
        Raises:
            TypeError: arabic does not satisfy  :code:`isinstance(arabic, numbers.Integral)` must be true.
            ValueError: arabic does not satisfy :code:`1 <= v <= 3999`

        Returns:
            str : string encoding the input as Roman numeral
        """

        if not isinstance(arabic,  numbers.Integral):
            raise TypeError("expected integer, got {}".format(type(arabic)))

        if arabic < 1 or arabic > 3999:
            raise ValueError("Argument must be between 1 and 3999")
        
        result = []
        for val, chars in ArabicToRoman._int_nums:
            count = int(arabic // val)
            result.append( chars * count)
            arabic -= val * count

        return ''.join(result) 


def arabic_to_roman(arabic : int) -> str:
    """Convert an Arabic numeral to a Roman numeral

    Shorthand for :py:meth:`ArabicToRoman.convert`, see
    :py:meth:`ArabicToRoman.convert` for full documentation.
    
    Args:
        arabic: Arabic numeral represented as integer.
    
    Raises:
        TypeError: arabic does not satisfy  :code:`isinstance(arabic, numbers.Integral)` must be true.
        ValueError: arabic does not satisfy :code:`1 <= v <= 3999`

    Returns:
        str : string encoding the input as Roman numeral
    """

    return ArabicToRoman.convert(arabic)